### PR TITLE
jefferson: init at 0.4.2

### PR DESCRIPTION
**Pull request by tnias**
###### Description of changes

Add jefferson (and cstruct). A JFFS2 filesystem extraction tool.

It can be used with binwalk or unblob.
For the latter there is an open package request https://github.com/NixOS/nixpkgs/issues/217836

###### Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [23.05 Release Notes (or backporting 22.11 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2305-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

**Comment by vlaci** [comment]
v0.4.2 is available on [PyPi](https://pypi.org/project/jefferson) and the source is maintained in another [repository](https://github.com/onekey-sec/jefferson).

@qkaiser can we make this less confusing? I guess a README update in jefferson would help.

**Comment by QKaiser** [comment]
> v0.4.2 is available on [PyPi](https://pypi.org/project/jefferson) and the source is maintained in another [repository](https://github.com/onekey-sec/jefferson).
> 
> @QKaiser can we make this less confusing? I guess a README update in jefferson would help.

Done. Everything's explained in the README now.

File: pkgs/development/python-modules/cstruct/default.nix [hunk 1]
```diff
@@ -0,0 +1,33 @@
...
+  version = "5.2";
+
+  src = fetchFromGitHub {
+    owner = "andreax79";
+    repo = "python-cstruct";
+    rev = "v${version}";
+    sha256 = "sha256-Dwogf0mmxFyBV7tPsuKV6gMZLPSCm7YhzqgJNHpaPFA=";
+  };
+
+  nativeBuildInputs = [
```
Comment (vlaci) [inline pkgs/development/python-modules/cstruct/default.nix:18]: I guess this should be:

```suggestion
  nativeCheckInputs = [
```

File: pkgs/development/python-modules/jefferson/default.nix [hunk 2]
```diff
@@ -0,0 +1,30 @@
...
+    owner = "sviehb";
+    repo = pname;
+    rev = "v${version}";
+    sha256 = "sha256-AUFOA3RuDQ57kt8WPQ4BQ/332n1dBvCvyD08PeX8+KI=";
+  };
+
+  propagatedBuildInputs = [
+    python-lzo
+    cstruct
+  ];
```
[the remark this answers is not in the document]
Comment (tnias) [reply pkgs/development/python-modules/jefferson/default.nix]: Did that and mark it as pyproject.

File: pkgs/development/python-modules/cstruct/default.nix (hunk 1 above)
Comment (tnias) [reply pkgs/development/python-modules/cstruct/default.nix]: Thanks, wasn't aware of nativeCheckInputs yet. :)

**Review by vlaci — APPROVED** [review]

**Comment by vlaci** [comment]
Thank you so much :tada:

**Comment by vlaci** [comment]
One thing that came to my mind is that jefferson is a CLI application in the first place, so `buildPythonApplication` could be a better fit and maybe relocating under `pkgs/tools/filesystems/` would be a better match. Do you have a library use-case for it?

Also It should be exposed in the top level IMHO.

I'll let the others chime in on that.

File: pkgs/development/python-modules/cstruct/default.nix [hunk 3]
```diff
@@ -0,0 +1,33 @@
...
+
+buildPythonPackage rec {
+  pname = "cstruct";
+  version = "5.2";
+
+  src = fetchFromGitHub {
+    owner = "andreax79";
+    repo = "python-cstruct";
+    rev = "v${version}";
+    sha256 = "sha256-Dwogf0mmxFyBV7tPsuKV6gMZLPSCm7YhzqgJNHpaPFA=";
```
Comment (dotlambda) [inline pkgs/development/python-modules/cstruct/default.nix:15]:
```suggestion
    hash = "sha256-Dwogf0mmxFyBV7tPsuKV6gMZLPSCm7YhzqgJNHpaPFA=";
```

File: pkgs/development/python-modules/cstruct/default.nix [hunk 4]
```diff
@@ -0,0 +1,33 @@
...
+  };
+
+  nativeCheckInputs = [
+    pytest
```
Comment (dotlambda) [inline pkgs/development/python-modules/cstruct/default.nix:19]:
```suggestion
    pytestCheckHook
```

File: pkgs/development/python-modules/cstruct/default.nix [hunk 5]
```diff
@@ -0,0 +1,33 @@
...
+  ];
+
+  checkPhase = ''
+    pytest
+  '';
```
Comment (dotlambda) [inline pkgs/development/python-modules/cstruct/default.nix:24]:
```suggestion
```

File: pkgs/development/python-modules/cstruct/default.nix [hunk 6]
```diff
@@ -0,0 +1,33 @@
+{ lib
+, buildPythonPackage
+, fetchFromGitHub
+, pytest
+}:
+
+buildPythonPackage rec {
+  pname = "cstruct";
+  version = "5.2";
```
Comment (dotlambda) [inline pkgs/development/python-modules/cstruct/default.nix:9]:
```suggestion
  version = "5.2";
  format = "setuptools";
```

File: pkgs/development/python-modules/jefferson/default.nix [hunk 7]
```diff
@@ -0,0 +1,38 @@
...
+buildPythonPackage rec {
+  pname = "jefferson";
+  version = "0.4.2";
+  format = "pyproject";
+
+  src = fetchFromGitHub {
+    owner = "onekey-sec";
+    repo = pname;
+    rev = "v${version}";
+    sha256 = "sha256-zW38vcDw4Jz5gO9IHrWRlvUznKvUyPbxkYMxn7VSTpA=";
```
Comment (dotlambda) [inline pkgs/development/python-modules/jefferson/default.nix:19]:
```suggestion
    hash = "sha256-zW38vcDw4Jz5gO9IHrWRlvUznKvUyPbxkYMxn7VSTpA=";
```

File: pkgs/development/python-modules/jefferson/default.nix [hunk 8]
```diff
@@ -0,0 +1,38 @@
...
+  nativeBuildInputs = [
+    poetry-core
+  ];
+
+  propagatedBuildInputs = [
+    click
+    cstruct
+    python-lzo
+  ];
+
```
Comment (dotlambda) [inline pkgs/development/python-modules/jefferson/default.nix:31]:
```suggestion

  # upstream has no tests
  doCheck = false;

```

**Review by dotlambda — CHANGES_REQUESTED** [review]
Please set `pythonImportsCheck`.

File: pkgs/development/python-modules/jefferson/default.nix (hunk 7 above)
Comment (tnias) [reply pkgs/development/python-modules/jefferson/default.nix]: applied

File: pkgs/development/python-modules/cstruct/default.nix (hunk 6 above)
Comment (tnias) [reply pkgs/development/python-modules/cstruct/default.nix:9]: applied

File: pkgs/development/python-modules/cstruct/default.nix (hunk 5 above)
Comment (tnias) [reply pkgs/development/python-modules/cstruct/default.nix]: applied

File: pkgs/development/python-modules/cstruct/default.nix (hunk 4 above)
Comment (tnias) [reply pkgs/development/python-modules/cstruct/default.nix]: applied

File: pkgs/development/python-modules/cstruct/default.nix (hunk 3 above)
Comment (tnias) [reply pkgs/development/python-modules/cstruct/default.nix]: applied

File: pkgs/development/python-modules/jefferson/default.nix (hunk 8 above)
Comment (tnias) [reply pkgs/development/python-modules/jefferson/default.nix:31]: applied

**Comment by vlaci** [comment]
> One thing that came to my mind is that jefferson is a CLI application in the first place, so `buildPythonApplication` could be a better fit and maybe relocating under `pkgs/tools/filesystems/` would be a better match. Do you have a library use-case for it?
> 
> Also It should be exposed in the top level IMHO.
> 
> I'll let the others chime in on that.

@tnias this question still stands

**Comment by tnias** [comment]
@vlaci haven't forgotten you, I was reading docs and trying to figure out what to do. :)

I don't have a library usecase for jefferson. I only know of binwalk and unblob, which both just call the executable. And jefferson README calls it a tool.

Plus the manual states:
> [...] The main purpose of this function [(buildPythonApplication)] is to build a Python package where one is interested only in the executables, and not importable module. [...]

=> I will convert it to application, like you suggested.

**Comment by tnias** [comment]
I think I have incooperated all feedback. Let me know, if I missed something or if there are any other wishes. :smile:

**Review by vlaci — APPROVED** [review]

**Review by dotlambda — APPROVED** [review]

**Pull request merged**